aten/src/ATen/core/intrusive_ptr.h

```diff
@@ -0,0 +1,532 @@
+        "Type mismatch. weak_intrusive_ptr move constructor got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. weak_intrusive_ptr move constructor got pointer with differing null value.");
+    rhs.target_ = FromNullType::singleton();
+  }
+
+  weak_intrusive_ptr(const weak_intrusive_ptr& rhs) noexcept
+      : target_(rhs.target_) {
+    retain();
+  }
+
+  template <class From, class FromNullType>
+  weak_intrusive_ptr(const weak_intrusive_ptr<From, FromNullType>& rhs) noexcept
+      : target_(rhs.target_) {
+    static_assert(
+        std::is_convertible<From*, TTarget*>::value,
+        "Type mismatch. weak_intrusive_ptr copy constructor got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. weak_intrusive_ptr copy constructor got pointer with differing null value.");
+    retain();
+  }
+
+  ~weak_intrusive_ptr() noexcept {
+    release();
+  }
+
+  weak_intrusive_ptr& operator=(weak_intrusive_ptr&& rhs) & noexcept {
+    return operator=<TTarget, NullType>(std::move(rhs));
+  }
+
+  template <class From, class FromNullType>
+      weak_intrusive_ptr& operator=(
+          weak_intrusive_ptr<From, FromNullType>&& rhs) &
+      noexcept {
+    static_assert(
+        std::is_convertible<From*, TTarget*>::value,
+        "Type mismatch. weak_intrusive_ptr move assignment got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. weak_intrusive_ptr move assignment got pointer with differing null value.");
+    release();
+    target_ = rhs.target_;
+    rhs.target_ = FromNullType::singleton();
+    return *this;
+  }
+
+  weak_intrusive_ptr& operator=(const weak_intrusive_ptr& rhs) & noexcept {
+    return operator=<TTarget, NullType>(rhs);
+  }
+
+  template <class From, class FromNullType>
+      weak_intrusive_ptr& operator=(
+          const weak_intrusive_ptr<From, NullType>& rhs) &
+      noexcept {
+    static_assert(
+        std::is_convertible<From*, TTarget*>::value,
+        "Type mismatch. weak_intrusive_ptr copy assignment got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. weak_intrusive_ptr copy assignment got pointer with differing null value.");
+    release();
+    target_ = rhs.target_;
+    retain();
+    return *this;
+  }
+
+  void reset() noexcept {
+    release();
+  }
+
+  void swap(weak_intrusive_ptr& rhs) noexcept {
+    TTarget* tmp = target_;
+    target_ = rhs.target_;
+    rhs.target_ = tmp;
+  }
+
+  size_t use_count() const noexcept {
+    if (target_ == NullType::singleton()) {
+      return 0;
+    }
+    return target_->refcount_.load(); // refcount, not weakcount!
+  }
+
+  bool expired() const noexcept {
+    return use_count() == 0;
```

Comment:
This will report that the NullType::singleton() is expired. I... don't think this is what you want?

Reply:
isn't it? I think that's what semantically makes sense. Does it hurt one of our use cases?